src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -678,204 +647,109 @@ object LazyList extends LazyListFactory[LazyList] {
-    case _ => fromIterator(coll.iterator)
-  }
+  def from(start: Int): LazyList[Int] = from(start, 1)
 
   /**
-    * @return A `Stream[A]` that gets its elements from the given `Iterator`.
+    * Create an infinite LazyList containing the given element expression (which
+    * is computed for each occurrence).
     *
-    * @param it Source iterator
-    * @tparam A type of elements
+    * @param elem the element composing the resulting LazyList
+    * @return the LazyList containing an infinite number of elem
     */
-  // Note that the resulting `Stream` will be effectively iterable more than once because
-  // `Stream` memoizes its elements
-  def fromIterator[A](it: Iterator[A]): Stream[A] =
-    if (it.hasNext) {
-      new Stream.Cons(it.next(), fromIterator(it))
-    } else Stream.Empty
+  def continually[A](elem: => A): LazyList[A] = newLL(sCons(elem, continually(elem)))
+
+  // significantly simpler than the iterator returned by Iterator.unfold
+  override def unfold[A, S](init: S)(f: S => Option[(A, S)]): LazyList[A] =
+    newLL {
+      f(init) match {
+        case Some((elem, state)) => sCons(elem, unfold(state)(f))
+        case None                => State.Empty
+      }
+    }
 
-  def empty[A]: Stream[A] = Empty
+  def newBuilder[A]: Builder[A, LazyList[A]] = ArrayBuffer.newBuilder[A].mapResult(array => from(array))
 
   // scalac generates a `readReplace` method to discard the deserialized state (see https://github.com/scala/bug/issues/10412).
   // This prevents it from serializing it in the first place:
   private[this] def writeObject(out: ObjectOutputStream): Unit = ()
   private[this] def readObject(in: ObjectInputStream): Unit = ()
+
+  /** This serialization proxy is used for LazyLists and Streams which start with a sequence of evaluated cons cells.
+    * The forced sequence is serialized in a compact, sequential format, followed by the unevaluated tail, which uses
+    * standard Java serialization to store the complete structure of unevaluated thunks. This allows the serialization
+    * of long evaluated streams without exhausting the stack through recursive serialization of cons cells.
+    */
+  @SerialVersionUID(3L)
+  class SerializationProxy[A](@transient protected var coll: LazyList[A]) extends Serializable {
+
+    private[this] def writeObject(out: ObjectOutputStream): Unit = {
+      out.defaultWriteObject()
+      var these = coll
+      while(these.headDefined) {
+        out.writeObject(these.head)
+        these = these.tail
+      }
+      out.writeObject(SerializeEnd)
+      out.writeObject(these)
+    }
+
+    private[this] def readObject(in: ObjectInputStream): Unit = {
+      in.defaultReadObject()
+      val init = new ArrayBuffer[A]
+      var initRead = false
+      while (!initRead) in.readObject match {
+        case SerializeEnd => initRead = true
+        case a => init += a.asInstanceOf[A]
+      }
+      val tail = in.readObject().asInstanceOf[LazyList[A]]
```

Comment:
This line throws a CCE (`LazyList.SerializationProxy` is not a `LazyList`), and I'm not sure why. Presumably it's a problem with the `writeObject` method, but I don't know what.

Reply:
Was a logic error in `writeReplace()` (as well as `State` not being `Serializable`